File: T1.SqlSharp/T1.SqlSharp/Expressions/SqlOverOrderByClause.cs
using T1.Standard.IO;

namespace T1.SqlSharp.Expressions;

public class SqlOverOrderByClause : ISqlExpression
{
    public SqlType SqlType { get; set; } = SqlType.OverOrderBy;
    public TextSpan Span { get; set; } = new();
    public void Accept(SqlVisitor visitor)
    {
        visitor.Visit_OverOrderByClause(this);
    }

    public ISqlExpression Field { get; set; } = new SqlFieldExpr();
    public List<SqlOrderColumn> Columns { get; set; } = [];

    public string ToSql()
    {
        var sql = new IndentStringBuilder();
        sql.Write(Field.ToSql());
        sql.Write(" OVER (");
        if (Columns.Count > 0)
        {
            sql.Write("ORDER BY ");
            foreach (var column in Columns.Select((value, index) => new { value, index }))
            {
                sql.Write(column.value.ToSql());
                if (column.index < Columns.Count - 1)
                {
                    sql.Write(", ");
                }
            }
        }
        sql.Write(")");
        return sql.ToString();
    }
}

File: VimSharp/chrome-extension-sharp/src/t1-sqlts/Expressions/SqlOverOrderByClause.ts
import { SqlType } from './SqlType';
import { TextSpan } from '../StringParser';
import { ISqlExpression, SqlVisitor } from './ISqlExpression';
import { SqlOrderColumn } from './SqlOrderColumn';
import { SqlFieldExpr } from './SqlFieldExpr';

export class SqlOverOrderByClause implements ISqlExpression {
    SqlType: SqlType = SqlType.OverOrderBy;
    Span: TextSpan = new TextSpan();
    Field: ISqlExpression = new SqlFieldExpr();
    Columns: SqlOrderColumn[] = [];

    Accept(visitor: SqlVisitor): void {
        visitor.Visit_OverOrderByClause(this);
    }

    ToSql(): string {
        let sql = this.Field.ToSql() + ' OVER (';
        if (this.Columns.length > 0) {
            sql += 'ORDER BY ' + this.Columns.map((col) => col.ToSql()).join(', ');
        }
        sql += ')';
        return sql;
    }
} 